### ISSUE TITLE
Stream.conMapJob

### ISSUE THREAD

**Issue by vasily-kirichenko** [Hopac/Hopac#93]
I'm trying to implement a combinator which maps stream in up to given number of jobs:
```fsharp
module Stream =
    let conMapJob (jobCount: int) (f: 'x -> Job<'y>) (s: Stream<'x>) : Stream<'y> =
        let inCh = Ch()
        let outCh = Ch()
        job {
            do! Stream.iterJob (Ch.give inCh) s |> Job.start
            do! Job.forN jobCount ((inCh >>= f >>= Ch.give outCh) |> Job.start)
        } |> start
        Stream.indefinitely outCh


Stream.unfoldFun (fun i -> Some(i, i + 1)) 0
|> Stream.take 1000L
|> Stream.conMapJob 2 (fun x ->
    job {
        printfn "in %d" x
        do! timeOutMillis 1000
        return x
    })
|> Stream.iterFun (printfn "out %d")
|> run
```
The output is:
```
in 0
in 1
out 0
out 1
```
After which the stream seems to "stop". What am I doing wrong?

**Comment by vasily-kirichenko** [comment]
Found the bug: instead of `do! Job.forN jobCount ((inCh >>= f >>= Ch.give outCh) |> Job.start)` it must be `do! Job.forN jobCount ((inCh >>= f >>= Ch.give outCh) |> Job.foreverServer)`.

**Comment by vasily-kirichenko** [comment]
It turns out it blocks forever on `run` because `Stream.indefinitely` never returns. The fixed version:
```fsharp
let conMapJob (jobCount: int) (f: 'x -> Job<'y>) (s: Stream<'x>) : Stream<'y> =
    let inCh = Ch()
    let inStreamClosed = IVar()
    let outCh = Ch()
        
    job {
        do! Stream.iterJob (Ch.give inCh) s >>=. IVar.fill inStreamClosed () |> Job.start
        return! Job.forN jobCount ((inCh >>= f >>= Ch.give outCh) |> Job.foreverServer)
    } |> start

    Stream.unfoldJob (fun _ -> 
        outCh ^-> fun x -> Some(x, ()) 
        <|>
        inStreamClosed ^-> fun _ -> None
    ) ()
```
@polytypic does it look good to you?

**Comment by polytypic** [comment]
@vasily-kirichenko Sorry, I didn't have time to look into this more carefully earlier. Here is the `searchWithHopac` function from the [gist](https://gist.github.com/vasily-kirichenko/626f570188ba26921e86fc013a583b55):

```fs
let searchWithHopac (file: string) =
    let reader = new StreamReader(file)
    Stream.indefinitely (job { return! reader.ReadLineAsync() })
    |> Stream.onCloseFun reader.Dispose
    |> Stream.take 1000000L
    |> Stream.conMapFun 4 (fun line ->
         match regex.Match line with
         | m when m.Success -> Some m.Groups.["name"].Value
         | _ -> None)
    |> Stream.chooseFun id
    |> Stream.conMapFun 4 (fun name -> job { return normalizeName name })
    |> Stream.count
```

There are some issues in the above.

The second use of `Stream.conMapFun` seems incorrect, because `Stream.conMapFun` takes a function of type `'x -> 'y` and is given a function of type `_ -> Job<_>`.

The use of `onCloseFun` has no effect.  As described in the documentation of [`onCloseFun`](http://hopac.github.io/Hopac/Hopac.html#def:val Hopac.Stream.onCloseFun) the given function is called when the *given* stream is closed.  To put it another way, the given function is not called when just the *returned* stream is closed.  You could fix this by moving the `onCloseFun` after `take`:

```diff
    Stream.indefinitely (job { return! reader.ReadLineAsync() })
-   |> Stream.onCloseFun reader.Dispose
-   |> Stream.take 1000000L
+   |> Stream.take 1000000L
+   |> Stream.onCloseFun reader.Dispose
    |> Stream.conMapFun 4 (fun line ->
```

However, in this case there is a better solution. Namely, you can use [`Job.using`](http://hopac.github.io/Hopac/Hopac.html#def:val Hopac.Job.using) and that is better for two reasons:
* It is faster, because it moves the finalization out of the stream expression.
* It does what the other benchmarks also do.  IOW, having finalization in the Hopac stream makes it different from the other benchmarks (Akka and AsyncSeq).

So, here is a rewritten version:

```fs
  let searchWithHopac (file: string) =
    Job.using <| new StreamReader(file) <| fun reader ->
    Stream.indefinitely <| job { return! reader.ReadLineAsync() }
    |> Stream.take 400000L
    |> Stream.conMapFun 4
        (regex.Match >> function
          | m when m.Success -> Some m.Groups.["name"].Value
          | _ -> None)
    |> Stream.choose
    |> Stream.conMapFun 4 normalizeName
    |> Stream.count
```

Now, of course, the main issue in all of this is that Hopac streams are not really a particularly good construct for *parallel* processing and the same can be said of AsyncSeq and Akka streams.  One problem is that streams imply a one-at-a-time model of processing.  Almost all of the combinators in the above pipeline, such as `indefinitely`, `take`, `choose` and `count` are sequential one-at-a-time operations.  They are all sequential bottlenecks.  Unless the operations done by the `conMapFun` blocks are really expensive, there will be no parallelism whatsoever.  If you replace `conMapFun` calls with `mapFun` calls, the stream might even run faster and will definitely take less total CPU time.

To get parallel speed ups, you must make a fundamental move from one-at-a-time to many-at-a-time such that there are no sequential bottlenecks in the pipeline.

**Comment by polytypic** [comment]
@vasily-kirichenko Here is the `conMapFun` function from the [gist](https://gist.github.com/vasily-kirichenko/626f570188ba26921e86fc013a583b55):

```fs
module Stream =
    let conMapFun (count: int) (f: 'x -> 'y) (s: Stream<'x>) : Stream<'y> =
        let inCh = Ch()
        let inStreamClosed = IVar()
        let outCh = Ch()
    
        job {
            do! Stream.iterJob (Ch.give inCh) s >>=. IVar.fill inStreamClosed () |> Job.start
            return! Job.forN count ((inCh >>- f >>= Ch.give outCh) |> Job.foreverServer)
        } |> start
    
        Stream.unfoldJob (fun _ -> 
            outCh ^-> fun x -> Some(x, ()) 
            <|>
            inStreamClosed ^-> fun _ -> None
        ) ()
```

There are some issues and things to note in the above.

First a note on behavior.  `conMapFun` operates just slightly differently from *most* of the stream functions provided by Hopac in that `conMapFun` is eagerish.  Most (but not all) of the Hopac stream functions have been constructed in such a way that they are lazy: *nothing* is done unless the head of the returned stream is actually requested.  A notable exception is [`keePreceding1`](http://hopac.github.io/Hopac/Hopac.html#def:val Hopac.Stream.keepPreceding1), which eagerly consumes the given stream to return a lazy stream.

What the above `conMapFun` does is that it starts to consume the given stream eagerly.  However, it only consumes just roughly `count` elements from the given stream in advance, because it uses synchronous messaging in between and processing "servers" will block unless the returned stream is consumed, so it is not entirely eager.  That only a given maximum number of elements are processed eagerly in parallel seems like a good compromise to me, because, while possible, a fully lazy version of `conMapFun` would be more complex and would also be less effective, because laziness is fundamentally incompatible with parallelism.

One problem with the above `conMapFun` is that it has a race condition.  Consider the following:

```fs
            do! Stream.iterJob (Ch.give inCh) s >>=. IVar.fill inStreamClosed () |> Job.start
...
        Stream.unfoldJob (fun _ -> 
            outCh ^-> fun x -> Some(x, ()) 
            <|>
            inStreamClosed ^-> fun _ -> None
        ) ()
```

What happens is that as soon as the input stream is finished, the `inStreamClosed` variable is written to signal end of stream.  However, at that point there can be roughly `count` operations still being processed.  When that happens, the returned stream may end prematurely.  Results are not yet given on `outCh`, but the `inStreamClosed` alternative is available and therefore the unfold will be finished, and the remaining operations in-flight will be ignored.

Another issue in the above `conMapFun` operation is that it doesn't handle exceptions.  If the given function `f` raises an exception, then things will go wrong.  I would recommend making it so that at least one exception is reported to the stream consumer.  It is possible to make it so that stream processing is stopped and all (max `count`) exceptions are reported to the stream consumer (as an aggregate exception), but that is more complex to implement.

Here is rewritten version of `conMapFun` that fixes the issues:

```fs
    let inline (>=>.) a b = a >=> fun _ -> b

    let rec fromUntil ch event =
           ch ^-> fun x -> Cons (x, fromUntil ch event)
      <|>* event

    let conMapFun degree x2y xs =
      let inn = Ch()
      let finish = IVar()
      let allDone = Latch.Now.create 1
      let dec = Latch.decrement allDone
      let out = Ch()
 
      xs
      |> Stream.iterJob ^ fun x ->
            Latch.Now.increment allDone
            Ch.give inn x
      >>=. dec
      |> start

      Job.tryWith
          (inn >>= (x2y >> Ch.give out >=>. dec) |> Job.forever)
          (IVar.tryFillFailure finish >=> Job.abort)
      |> Job.server
      |> Job.forN degree
      |> start

      allDone ^=>. IVar.tryFill finish Nil |> start

      fromUntil out finish
```

Note that the above uses `IVar.tryFillFailure` which is not available in the currently released version of Hopac&mdash;I just [added it to the master branch](https://github.com/Hopac/Hopac/commit/5f9971271c502f8960a6c62ea75df4bbb64386ed). To compile the code with the released Hopac version you could replace it `IVar.fillFailure`, but that will not work correctly in case exceptions are actually raised, because it might lead to attempts to write to the `IVar` multiple times.

**Comment by vasily-kirichenko** [comment]
@polytypic Thanks for fixing and explanation! I updated the blog post.
